### ISSUE TITLE
page: tag page changes to common presentational elements

### ISSUE THREAD

**Issue by Bloke** [textpattern/textpattern.github.io#155]
### Absolute URL of front-end page

(e.g.) https://docs.textpattern.com/tags/images

### The question/problem

We have the notion of 'global' attributes now that all tags can take. Instead of repeating the definition of these ad nauseum in each tag page, can we use includes?

#### Option A
Write a whole bunch of includes for each set of things we use often, e.g. a doc that has the heading, intro text, plus class, wraptag, break, escape in it. And a doc that has heading, intro text, plus class, wraptag, break, escape, breakby, label, labeltag. And so on.

#### Option B (better)
Write a bunch of individual includes, one for each attribute such as escape, break, label, labeltag, breakby, wraptag, class. Then, on each tag page under the 'Common attributes' heading, call them into the doc one by one on a tag-by-tag basis.

#### Option C (better still?)
Write a single common presentational attributes header include file with the header text, introductory paragraph and then a bunch of conditionals in it, one for each attribute. When you call in the include, you pass in the attributes you want to display and the include only displays the ones you give it in the parameter list.

e.g. if using:

`{% include common_atts.html class breakby wraptag %}`

the include only displays those three under the heading and intro text.

I'm not sure how capable includes and parameters are. I've not seen any examples yet where you can pass in valueless attributes so we might have to do something stupid like:

`{% include common_atts.html class="1" breakby="1" wraptag="1" %}`

and then check to see if each common attribute is equal to 1 in order to display it or not. BUt the general idea is that we write a single template with a bunch of conditionals and then just render bits of it based on the way that file is included and the params we give it.

Worth pursuing?

**Comment by Bloke** [comment]
Another avenue for Option C might be to pass in something like this:

`{% include common_atts.html atts="class, breakby, wraptag" %}`

and then in the include iterate over them:

```
{% for att in include.atts %}
   {% if {{att}} = "class" %}
      // display class attribute info
   {% endif %}
   {% if {{att}} = "breakby" %}
      // display breakby attribute info
   {% endif %}
   ...
{% endfor %}
```

**Comment by wion** [comment]
I think anytime you can eliminate overhead editing by single-sourcing... Do it. And the tag pages seem to be prime candidates for simplified editing wins wherever they can be gained. Especially as there are so damn many of them.

As for understanding the liquid options and whatnot, thus what is the best course, I'll have to defer to your and Phil's better judgement on how to proceed there. I get lost when anything is too deeply woven around conditionals, and liquid has always been a bit abstract to me.

At any rate, I'm licking my wounds and directing my attention away from tags pages now, back to the other site sections and needed materials, and the various guidelines I still need to finish.

**Comment by philwareham** [comment]
Yes we should always try to utilise reusable code wherever possible to make everyones life easier. So I'm in favour of includes wherever we can.

**Comment by Bloke** [comment]
I'll see what I can do.

**Comment by Bloke** [comment]
Okay, it works. I've added an '_includes/common-atts.html' file and have coded in entries so far for:

* break
* breakby
* class
* escape (global)
* label
* labeltag
* style
* wraptag

We can add more entries, plus the additional global attributes, as and when we need them.

To use it in a tag page, call the include file and pass it a list of attributes, using the attribute name and then its default value as the argument. That value will be used verbatim in the tag default text, which means we can get rid of the silly 'see exceptions' thing, as we can put the exact defaults in per tag.

So, for the [section_list](https://docs.textpattern.com/tags/section_list) tag:

```
{% include common-atts.html break="br" breakby="unset" class="tag name or unset" escape="html" label="unset" labeltag="unset" wraptag="unset" %}
```

That's live for that tag now. I've added automatic links to the cross-reference pages for each one, noted if an attribute is global or not, and also (for escape) linked to the tag-escaping doc, which will move when the tag basics area is consolidated. But once all tag pages are converted to use this, we'll only need to update the link once in the include file. Woohoo!

Any amendments, improvements and help converting the tag pages over to use it welcome.

**Comment by Bloke** [comment]
Made some more tweaks to the include file so you don't need to specify the default value if it's the most common default we use. Unfortunately, Liquid doesn't accept valueless attributes so if you want to use the default value you specify `=""`. The defaults are:

* break (br)
* breakby (unset)
* class (unset)
* escape (html)
* label (unset)
* labeltag (unset)
* style (unset)
* wraptag (unset)

So the above mentioned include for section_list could now be:

```
{% include common-atts.html break="" breakby="" class="tag name or unset" escape="" label="" labeltag="" wraptag="" %}
```

**Comment by Bloke** [comment]
Moved the global atts out of the common-atts.html file into their own [global-atts include file](#157).

So for the above mentioned section_list tag page, the include now becomes:

```
{% include common-atts.html break="" breakby="" %}
```

**Comment by wion** [comment]
> Any amendments, improvements and help converting the tag pages over to use it welcome.

I've moved the tag pages checklist from #153 to head post here, to help keep track. You might want to added to the change list just above, as needed.

**Comment by wion** [comment]
#127

**Comment by wion** [comment]
#118

**Comment by Bloke** [comment]
For completeness, I've renamed the include file so it is grouped better in the file list with the global attributes include:

```
{% include atts-common.html %}
```
